serlcd/serlcd.go

```diff
@@ -0,0 +1,249 @@
+// Cursor(CursorOff, CursorUnderline)
+func (dev *Dev) Cursor(mode ...display.CursorMode) (err error) {
+	dev.displayDCB &= 0x04
+	for _, cmd := range mode {
+		switch cmd {
+		case display.CursorBlink:
+			dev.displayDCB |= 0x01
+		case display.CursorUnderline:
+			dev.displayDCB |= 0x02
+		case display.CursorBlock:
+			dev.displayDCB |= 0x01
+		case display.CursorOff:
+		default:
+			err = wrap(display.ErrInvalidCommand)
+			return
+		}
+	}
+	dev.displayDCB = (dev.displayDCB | 0x08) & 0xf
+
+	_, err = dev.Write([]byte{cmdMode, dev.displayDCB})
+	return
+}
+
+// Halt shuts down the display. If the IO source implements io.Closer, it is
+// called.
+func (dev *Dev) Halt() (err error) {
+	err = dev.Clear()
+	if err != nil {
+		return
+	}
+	err = dev.Display(false)
+	if err != nil {
+		return
+	}
+	if dev.w != nil {
+		if cl, ok := dev.w.(io.Closer); ok {
+			err = cl.Close()
+		}
+	}
+	return
+}
+
+// Move the cursor home (MinRow(),MinCol())
+func (dev *Dev) Home() (err error) {
+	err = dev.MoveTo(dev.MinRow(), dev.MinCol())
+	time.Sleep(2 * time.Millisecond)
+	return
+}
+
+// Return the min column position.
+func (dev *Dev) MinCol() int {
+	return 0
+}
+
+// Return the min row position.
+func (dev *Dev) MinRow() int {
+	return 0
+}
+
+// Move the cursor forward or backward.
+func (dev *Dev) Move(dir display.CursorDirection) (err error) {
+	cmdByte := byte(0x10)
+	switch dir {
+	case display.Backward:
+		// Nothing
+	case display.Forward:
+		cmdByte |= 0x04
+	case display.Down:
+		fallthrough
+	case display.Up:
+		fallthrough
+	default:
+		err = wrap(display.ErrNotImplemented)
+		return
+	}
+	_, err = dev.Write([]byte{cmdMode, cmdByte})
+	return
+}
+
+// Move the cursor to an arbitrary position.
+func (dev *Dev) MoveTo(row, col int) (err error) {
+	lineOffsets := []byte{0, 64, 20, 84}
+	if row < dev.MinRow() || row >= dev.Rows() ||
+		col < dev.MinCol() || col >= dev.Cols() {
+		return errors.New("serlcd: invalid MoveTo() offset")
+	}
+	cmdByte := byte(0x80) + lineOffsets[row] + byte(col)
+	_, err = dev.Write([]byte{cmdMode, byte(cmdByte)})
+	return
+}
+
+// Return the number of rows the display supports.
+func (dev *Dev) Rows() int {
+	return dev.rows
+}
+
+// Turn the display on / off
+func (dev *Dev) Display(on bool) (err error) {
+	if on {
+		dev.displayDCB |= 0x04
+	} else {
+		dev.displayDCB ^= 0x04
+	}
+	_, err = dev.Write([]byte{cmdMode, (dev.displayDCB | 0x08) & 0x0f})
+	return
+}
+
+// return info about the display.
+func (dev *Dev) String() string {
+	ioType := "None"
+	if dev.conn != nil {
+		ioType = "periph.io.Conn.Conn"
+	} else if dev.w != nil {
+		ioType = fmt.Sprintf("%#v", dev.w)
+	}
+	return fmt.Sprintf("SparkFun SerLCD %dx%d Display - %s", dev.cols, dev.rows, ioType)
+}
+
+// Write a set of bytes to the display.
+func (dev *Dev) Write(p []byte) (n int, err error) {
+	if dev.w != nil {
+		n, err = dev.w.Write(p)
+		return
+	}
+	// Evidently, for i2c there's a buffer limitation of 32 bytes. Writing
```

Comment:
Isn't that device specific? It's okay here but please mention "Device XYZ has a write buffer of 32 bytes. Make it configurable in the future if needed."

Reply:
This is a really specific device. They've got a board with an Atmel processor on it running a custom arduino image and that talks to the display, and drives the LEDs for the RGB. There's no way to make this more generic than having the interface that it complies with.

Reply:
An alternative is to have the New**() functions to accept an Options struct, so that this could become configurable in the future as a non-breaking change? For now, the Options struct would contain cols and rows.

Reply:
I've done a little work on this. I've included a link to the SparkFun issue about the 32 byte limit. The issue is 5 years old, and doesn't have a resolution and from the tone, it never will. 

This isn't an actively developed product category, so I think adding complexity for future compatibility is just adding complexity. I've restructured the 32 char limit to make it more transparent what's going on should it change in the future.